components/brave_rewards/resources/page/reducers/promotion_reducer.ts

```diff
@@ -0,0 +1,234 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Reducer } from 'redux'
+
+// Constant
+// Temporary (ryanml)
+import { types } from '../constants/rewards_types'
+
+const getPromotion = (id: string, promotions?: Rewards.Promotion[]) => {
+  if (!promotions) {
+    return undefined
+  }
+
+  return promotions.find((promotion: Rewards.Promotion) => {
+    return (promotion.promotionId === id)
+  })
+}
+
+const updatePromotions = (newPromotion: Rewards.Promotion, promotions: Rewards.Promotion[]): Rewards.Promotion[] => {
+  return promotions.map((promotion: Rewards.Promotion) => {
+    if (newPromotion.promotionId === promotion.promotionId) {
+      return Object.assign(promotion, newPromotion)
+    }
+    return promotion
+  })
+}
+
+const updatePromotion = (newPromotion: Rewards.Promotion, promotions: Rewards.Promotion[]): Rewards.Promotion => {
+  const oldPromotion = promotions.filter((promotion: Rewards.Promotion) => newPromotion.promotionId === promotion.promotionId)
+
+  if (oldPromotion.length === 0) {
+    return newPromotion
+  }
+
+  return Object.assign(oldPromotion[0], newPromotion)
+}
+
+const promotionReducer: Reducer<Rewards.State | undefined> = (state: Rewards.State, action) => {
+  const payload = action.payload
+  switch (action.type) {
+    case types.FETCH_PROMOTIONS: {
+      chrome.send('brave_rewards.fetchPromotions')
+      break
+    }
+    case types.ON_PROMOTIONS: {
+      state = { ...state }
+      if (payload.properties.result === 1) {
+        break
+      }
+
+      if (!state.promotions) {
+        state.promotions = []
+      }
+
+      let promotions = payload.properties.promotions
+
+      promotions = promotions.map((promotion: Rewards.Promotion) => {
+        promotion.expiresAt = promotion.expiresAt * 1000
+        return updatePromotion(promotion, state.promotions || [])
+      })
+
+      state = {
+        ...state,
```

Comment:
Not familiar with typescript`...state,` notation, what does this mean?

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

Reply:
aha - vargs to me :) thank you!